### PR TITLE
Visitor refactor

### PR DESCRIPTION
**Pull request by soyombo-baterdene**
[ISSUE](https://github.com/erxes/erxes/issues/2588)

### Context
Problem:
Each time someone visits a website that has a messenger integration connected, we saved that visitor's data in the main database. This made a lot of unnecessary data is stored in the customers collection.

Solution:
- Save the visitor data in a separate database. Update the main database when the visitor interacts through chat.
- Add custom command that cleans previous visitor data

### CODE REVIEW

File: api/src/__tests__/widgetMutations.test.ts [hunk 1]
```diff
@@ -636,6 +681,51 @@ describe('saveBrowserInfo()', () => {
...
+        rules: [
+          {
+            text: 'text',
+            kind: 'currentPageUrl',
+            condition: 'is',
+            value: '/page'
+          }
+        ]
+      },
+      kind: 'visitorAuto',
```
Comment (Buyantogtokh) [inline api/src/__tests__/widgetMutations.test.ts:712]: use already declared constant values

File: api/src/data/resolvers/mutations/widgets.ts [hunk 2]
```diff
@@ -511,19 +519,20 @@ const widgetMutations = {
       uiOptions: integration.uiOptions,
       languageCode: integration.languageCode,
       messengerData: await getMessengerData(integration),
-      customerId: customer._id,
+      customerId: customer ? customer?._id : null,
```
Comment (Buyantogtokh) [inline api/src/data/resolvers/mutations/widgets.ts:522]: remove optional ?

File: api/src/data/widgetUtils.ts [hunk 3]
```diff
@@ -7,43 +7,78 @@ import {
...
+  let visitor;
+
+  if (visitorId) {
+    visitor = await getVisitorLog(visitorId);
+    integrationId = visitor.integrationId;
+  }
 
   // Preventing from displaying non messenger integrations like form's messages
   // as last unread message
   const integration = await Integrations.findOne({
```
Comment (Buyantogtokh) [inline api/src/data/widgetUtils.ts:36]: use Integrations.getIntegration(selector) static function which checks existence of the integration inside

File: api/src/data/widgetUtils.ts [hunk 4]
```diff
@@ -7,43 +7,78 @@ import {
...
+  const query = customerId
+    ? { integrationId, customerId }
+    : { integrationId, visitorId };
+
+  const convs = await Conversations.find(query);
 
   return Messages.findOne(Conversations.widgetsUnreadMessagesQuery(convs));
 };
+
+export const createVisitorFromVisitorLog = async (visitorId: string) => {
```
Comment (Buyantogtokh) [inline api/src/data/widgetUtils.ts:66]: function name does not sound good

File: api/src/data/widgetUtils.ts [hunk 5]
```diff
@@ -7,43 +7,78 @@ import {
...
+  const doc = { state: 'visitor', ...visitor };
+  const customer = await Customers.createCustomer(doc);
+
+  await Messages.updateVisitorEngageMessages(visitorId, customer._id);
+  await Conversations.updateMany(
+    {
+      visitorId
+    },
+    { $set: { customerId: customer._id, visitorId: '' } },
+    { multi: true }
```
Comment (Buyantogtokh) [inline api/src/data/widgetUtils.ts:80]: updateMany does not need extra { multi: true } option

File: api/src/db/models/Brands.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { IIntegrationDocument } from './definitions/integrations';
 
 export interface IBrandModel extends Model<IBrandDocument> {
   getBrand(_id: string): IBrandDocument;
+  getBrandByCode(code: string): IBrandDocument;
```
Comment (Buyantogtokh) [inline api/src/db/models/Brands.ts:9]: almost the same as getBrand(). No need for such duplicated code. Add optional code param to getBrand()

File: api/src/db/models/ConversationMessages.ts [hunk 7]
```diff
@@ -201,6 +202,20 @@ export const loadClass = () => {
...
+      visitorId: string,
+      customerId: string
+    ) {
+      return Messages.updateMany(
+        {
+          visitorId,
+          engageData: { $exists: true }
+        },
+        { $set: { customerId, visitorId: '' } },
+        { multi: true }
```
Comment (Buyantogtokh) [inline api/src/db/models/ConversationMessages.ts:216]: remove

File: api/src/db/models/Engages.ts [hunk 8]
```diff
@@ -253,13 +262,13 @@ export const loadClass = () => {
...
-          !customerIds.includes(customer._id)
+          !customerIds.includes(customerObj._id)
         ) {
           continue;
         }
 
         const customersSelector = {
-          _id: customer._id,
+          _id: customerObj._id,
           state: { $ne: 'visitor' },
```
Comment (Buyantogtokh) [inline api/src/db/models/Engages.ts:272]: use constant

File: api/src/db/models/Engages.ts [hunk 9]
```diff
@@ -284,12 +296,12 @@ export const loadClass = () => {
...
-          customer._id,
-          browserInfo.url
-        );
+        const numberOfVisits = await getNumberOfVisits({
+          url: browserInfo.url,
+          visitorId: visitor ? visitor.visitorId : undefined,
+          customerId: customer ? customer._id : undefined
+        });
 
         const isPassedAllRules = await this.checkRules({
```
Comment (Buyantogtokh) [inline api/src/db/models/Engages.ts:306]: rename this var, like hasPassedAllRules or allRulesPassed

File: api/src/db/models/Engages.ts [hunk 10]
```diff
@@ -318,8 +330,9 @@ export const loadClass = () => {
 
           const conversationMessage = await this.createOrUpdateConversationAndMessages(
             {
-              customer,
-              integration,
+              customerId: customer?._id,
```
Comment (Buyantogtokh) [inline api/src/db/models/Engages.ts:333]: remove ?

File: api/src/db/models/Integrations.ts [hunk 11]
```diff
@@ -362,17 +362,17 @@ export const loadClass = () => {
...
       brandObject = false
     ) {
-      const brand = await Brands.findOne({ code: brandCode });
-
-      if (!brand) {
-        throw new Error('Brand not found');
-      }
+      const brand = await Brands.getBrandByCode(brandCode);
 
       const integration = await Integrations.findOne({
```
Comment (Buyantogtokh) [inline api/src/db/models/Integrations.ts:367]: use Integrations.getIntegration()

File: api/src/db/models/definitions/conversationMessages.ts [hunk 12]
```diff
@@ -91,6 +92,7 @@ export const messageSchema = new Schema({
   conversationId: field({ type: String, index: true }),
   internal: field({ type: Boolean, index: true }),
   customerId: field({ type: String, index: true }),
+  visitorId: field({ type: String, index: true }),
```
Comment (Buyantogtokh) [inline api/src/db/models/definitions/conversationMessages.ts:95]: add label to schema field declarations

File: api/src/db/models/definitions/conversations.ts [hunk 13]
```diff
@@ -53,6 +54,7 @@ export const conversationSchema = new Schema({
   content: field({ type: String, optional: true }),
   integrationId: field({ type: String, index: true }),
   customerId: field({ type: String }),
+  visitorId: field({ type: String, index: true }),
```
Comment (Buyantogtokh) [inline api/src/db/models/definitions/conversations.ts:57]: add label

File: logger/src/models/Visitors.ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+  }),
+  sessionCount: field({
+    type: Number,
+    label: 'Session count',
+    optional: true
+  }),
+  createdAt: field({ type: Date, default: Date.now })
+});
+
+export const loadLogClass = () => {
```
Comment (Buyantogtokh) [inline logger/src/models/Visitors.ts:103]: rename this func

File: logger/src/models/Visitors.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+      return Visitors.findOne({ visitorId: doc.visitorId });
+    }
+
+    public static async getVisitorLog(visitorId: string) {
+      const visitor = Visitors.findOne({ visitorId });
+
+      if (!visitor) {
+        throw new Error('Visitor not');
+      }
+      return Visitors.findOne({ visitorId });
```
Comment (Buyantogtokh) [inline logger/src/models/Visitors.ts:166]: don't fetch visitor again, use previous one

File: logger/src/models/Visitors.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+  sessionCount: field({
+    type: Number,
+    label: 'Session count',
+    optional: true
+  }),
+  createdAt: field({ type: Date, default: Date.now })
+});
+
+export const loadLogClass = () => {
+  class Visitor {
```
Comment (Buyantogtokh) [inline logger/src/models/Visitors.ts:104]: too many not found errors thrown

File: logger/src/models/Visitors.ts [hunk 17]
```diff
@@ -0,0 +1,171 @@
+import { Document, model, Model, Schema } from 'mongoose';
+
+/**
+ * Mongoose field options wrapper
+ * @param {Object} options Mongoose schema options
+ */
+export const field = options => {
```
Comment (Buyantogtokh) [inline logger/src/models/Visitors.ts:7]: remove duplicate declaration

**Comment by Buyantogtokh** [comment]
Remove visitor from logs after the visitor becomes a customer

File: logger/src/messageBroker.ts [hunk 18]
```diff
@@ -13,11 +14,44 @@ export const initBroker = async server => {
...
+    if (action === 'update') {
+      return Visitors.updateVisitorLog(data);
+    } else if (action === 'createOrUpdate') {
+      return Visitors.createOrUpdateVisitorLog(data);
+    } else if (action === 'remove') {
+      return Visitors.removeVisitorLog(data.visitorId);
+    }
+  });
+
+  consumeRPCQueue('rpc_queue:visitorLog', async parsedObject => {
```
Comment (batamar) [inline logger/src/messageBroker.ts:35]: why rpc

File: api/src/data/resolvers/mutations/widgets.ts [hunk 19]
```diff
@@ -422,51 +428,71 @@ const widgetMutations = {
...
+            customData
+          })
+        : await Customers.createMessengerCustomer({ doc, customData });
+    }
+
+    if (visitorId) {
+      const isRunning = await isLoggerRunning();
+
+      if (isRunning) {
+        await sendToVisitorLog(
```
Comment (batamar) [inline api/src/data/resolvers/mutations/widgets.ts:485]: why can not we catch the exception in sendToVisitorLog without calling isLoggerRunning. ?

**Review by batamar — COMMENTED** [review]
try without elasticsearch

File: api/src/data/resolvers/mutations/widgets.ts [hunk 20]
```diff
@@ -422,51 +428,71 @@ const widgetMutations = {
...
 
     const customData = data;
 
     // find brand
-    const brand = await getBrand(brandCode);
-
-    if (!brand) {
-      throw new Error('Brand not found');
-    }
+    const brand = await Brands.getBrand({ code: brandCode });
```
Comment (batamar) [inline api/src/data/resolvers/mutations/widgets.ts:438]: use getBrand

**Comment by batamar** [comment]
command to remove

File: api/src/__tests__/widgetMutations.test.ts [hunk 21]
```diff
@@ -173,6 +176,57 @@ describe('messenger connect', () => {
...
+    );
+
+    expect(response.customerId).toBeUndefined();
+    expect(response.visitorId).toBe('123');
+
+    logUtilsMock.restore();
+    mock.restore();
+  });
+
+  test('creates new customer when logger is not running', async () => {
```
Comment (Buyantogtokh) [inline api/src/__tests__/widgetMutations.test.ts:205]: where is the new customer checking code?

File: api/src/apolloClient.ts [hunk 22]
```diff
@@ -143,6 +147,9 @@ export const initApolloServer = async (app) => {
             webSocket.messengerData = parsedMessage.value;
 
             const customerId = webSocket.messengerData.customerId;
```
Comment (Buyantogtokh) [inline api/src/apolloClient.ts:149]: Can webSocket or webSocket.messengerData be undefined or null? If so, this line could throw error.

File: api/src/__tests__/widgetMutations.test.ts (hunk 21 above)
Comment (soyombo-baterdene) [reply api/src/__tests__/widgetMutations.test.ts:205]: there is a piece of code in line 224. if the customer is not created, there is no customerId

File: api/src/commands/removeVisitors.ts [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+import { stream } from '../data/bulkUtils';
+import { connect } from '../db/connection';
+import { Conversations, Customers } from '../db/models';
+
+dotenv.config();
+
+const command = async () => {
+  await connect();
+
+  const customerIds = await Customers.find({
```
Comment (Buyantogtokh) [inline api/src/commands/removeVisitors.ts:11]: use aggregate() instead of find()

File: logger/src/messageBroker.ts [hunk 24]
```diff
@@ -13,11 +14,44 @@ export const initBroker = async server => {
...
+  const { consumeQueue, consumeRPCQueue } = client;
 
   consumeQueue('putLog', async data => {
     await receivePutLogCommand(data);
   });
+
+  consumeQueue('visitorLog', async parsedObject => {
+    const { data, action } = parsedObject;
+
+    if (action === 'update') {
```
Comment (Buyantogtokh) [inline logger/src/messageBroker.ts:26]: use switch case

File: widgets/client/widgetUtils.ts [hunk 25]
```diff
@@ -116,3 +117,13 @@ export const setErxesProperty = (name: string, value: any) => {
...
 };
+
+export const getVisitorId = async () => {
+  const fp = await FingerprintJS.load();
+
+  // The FingerprintJS agent is ready.
+  const result = await fp.get();
+
+  // This is the visitor identifier:
+  return result.visitorId;
```
Comment (Buyantogtokh) [inline widgets/client/widgetUtils.ts:128]: Could there be an error thrown?

Comment (soyombo-baterdene) [reply widgets/client/widgetUtils.ts:128]: no it does not

**Pull request merged**